app/controllers/criteria_controller.rb

```diff
@@ -114,10 +114,21 @@ def update
   # Handles the drag/drop criteria sorting.
   def update_positions
     @assignment = Assignment.find(params[:assignment_id])
+    invalid_criteria = false
 
     ApplicationRecord.transaction do
-      params[:criterion].each_with_index do |id, index|
-        Criterion.update(id, position: index + 1) unless id.blank?
+      Criterion.transaction(requires_new: true) do
```

Comment:
It's good that you're using a transaction here, and in fact we don't need to have a nested one. The outer transaction is a bit too broad; so you can remove this line and replace `ApplicationRecord` with `Criterion` in the previous line.